slides/week4/week4.org

```diff
@@ -3,57 +3,336 @@
+  - be careful how much you allocate
+  - returns a pointer to the memory
+- To deallocate the memory you must use *delete*
+  - anytime memory is allocated you must deallocate it
+  - you pass the pointer to the memory you got from new
+  - not deallocating is the cause of memory leaks
+    - stay tuned for a demo
+
+** New
+#+BEGIN_SRC c++
+int main() {
+  double *a = new double(10.0);
+}
+#+END_SRC
+[[file:https://imgur.com/z3xGol7.png]]
+
+** Delete
+#+BEGIN_NOTES
+use the stuct defined in week4.h, prints when constructed and destructed.
+#+END_NOTES
+#+BEGIN_SRC c++
+int main() {
+  double *a = new double(10.0);
+  // ... legit code
+  delete a;
+}
+#+END_SRC
+[[file:https://imgur.com/A6a4bsz.png]]
+
+*** Segfault
+#+BEGIN_NOTES
+tell them what a segfault is
+#+END_NOTES
+#+BEGIN_SRC c++
+int main() {
+  double *a = new double(10.0);
+  // ... legit code
+  delete a;
+  // .. more code
+  *a = 20.0
+  // .. more code
+}
+#+END_SRC
+
 * Smart pointers
 #+BEGIN_NOTES
-Each type has an initialization function. Run through some basic examples here.
+Each type has an initialization function. Run through examples in 4 slides
 #+END_NOTES
 - Wrap a normal pointer to make it easier to manage the lifecycle of objects
 - You should generally use one of these two types instead of a normal pointer:
  - =shared_ptr=: multiple =shared_ptr= s can own a single object
  - =unique_ptr=: only one =unique_ptr= can own an object
 
+** =unique_ptr=
+#+BEGIN_NOTES
+object is a struct defined in week4.h. make sure to use it as an example
+#+END_NOTES
+#+BEGIN_SRC c++
+int main() {
+  unique_ptr<object> u_ptr = make_unique<object>();
+}
+#+END_SRC
+#+ATTR_HTML: :width 70%
+[[file:https://i.imgur.com/nK51rsP.png]]
+
+** =shared_ptr=
+#+BEGIN_NOTES
+object is a struct defined in week4.h. make sure to use it as an example
+#+END_NOTES
+#+BEGIN_SRC c++
+int main() {
+  shared_ptr<object> s_ptr1 = make_shared<object>();
+  // <--- HERE
+  {
+    shared_ptr<object> s_ptr2 = make_shared<object>();
+  }
+  shared_ptr<object> s_ptr3 = make_shared<object>();
+}
+#+END_SRC
+#+ATTR_HTML: :width 70%
+[[file:https://i.imgur.com/Sy6CY5q.png]]
+
+*** =shared_ptr=
+#+BEGIN_NOTES
+object is a struct defined in week4.h. make sure to use it as an example
+#+END_NOTES
+#+BEGIN_SRC c++
+int main() {
+  shared_ptr<object> s_ptr1 = make_shared<object>();
+  {
+    shared_ptr<object> s_ptr2 = make_shared<object>();
+    // <--- HERE
+  }
+  shared_ptr<object> s_ptr3 = make_shared<object>();
+}
+#+END_SRC
+#+ATTR_HTML: :width 70%
+[[file:https://i.imgur.com/y3ITWBn.png]]
+
+*** =shared_ptr=
+#+BEGIN_NOTES
+object is a struct defined in week4.h. make sure to use it as an example
+#+END_NOTES
+#+BEGIN_SRC c++
+int main() {
+  shared_ptr<object> s_ptr1 = make_shared<object>();
+  {
+    shared_ptr<object> s_ptr2 = make_shared<object>();
+  }
+  shared_ptr<object> s_ptr3 = make_shared<object>();
+  // <--- HERE
+}
+#+END_SRC
+#+ATTR_HTML: :width 70%
+[[file:https://i.imgur.com/qLddS8b.png]]
```

Comment:
This image is misleading. `s_ptr2` get's deallocated from the stack (popped off), so `s_ptr3` would take it's place, not be placed after the `???`

Reply:
valid point. Changed graphic